qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+    and add parameters found to the array.
+    """
+    def __init__(self):
+        self.parameters = set()
+
+    # Top layer: Get parameters from programs
+
+    def visit_ScheduleBlock(self, node: ScheduleBlock):
+        """Visit ``ScheduleBlock``. Recursively visit context blocks and search parameters.
+
+        .. note:: ``ScheduleBlock`` can have parameters in blocks and its alignment.
+        """
+        for parameter in node.parameters:
+            self.parameters.add(parameter)
+
+    def visit_Schedule(self, node: Schedule):
+        """Visit ``Schedule``. Recursively visit schedule children and search parameters."""
+        for parameter in node.parameters:
+            self.parameters.add(parameter)
+
+    def visit_AlignmentKind(self, node: AlignmentKind):
+        """Get parameters from block's ``AlignmentKind`` specification."""
+        for param in node._context_params:
+            self.visit(param)
+
+    # Mid layer: Get parameters from instructions
+
+    def visit_Call(self, node: instructions.Call):
+        """Get parameters from ``Call`` instruction.
+
+        .. note:: ``Call`` instruction has a special parameter handling logic.
+            This instruction separately keeps parameters and program.
+        """
+        for parameter in node.parameters:
+            self.parameters.add(parameter)
+
+    def visit_Instruction(self, node: instructions.Instruction):
+        """Get parameters from general pulse instruction.
+
+        .. note:: All parametrized object should be stored in the operands.
+            Otherwise parameter cannot be detected.
+        """
+        for op in node.operands:
+            self.visit(op)
+
+    # Lower layer: Get parameters from operands
+
+    def visit_Channel(self, node: channels.Channel):
+        """Get parameters from ``Channel`` object."""
+        if isinstance(node.index, ParameterExpression):
+            self._parse_parameter_expression(node.index)
+
+    def visit_ParametricPulse(self, node: ParametricPulse):
+        """Get parameters from ``ParametricPulse`` object."""
+        for op_value in node.parameters.values():
+            if isinstance(op_value, ParameterExpression):
+                self._parse_parameter_expression(op_value)
+
+    def visit_Waveform(self, node: Waveform):
+        """Get parameters from ``Waveform`` object.
+
+        .. node:: No parameter can be assigned to ``Waveform`` object.
+        """
+        pass
+
+    def generic_visit(self, node: Any):
+        """Get parameters from object that doesn't belong to Qiskit Pulse module."""
+        if isinstance(node, ParameterExpression):
+            self._parse_parameter_expression(node)
+
+    def _parse_parameter_expression(self, param_expr: ParameterExpression):
+        """A helper function to get parameters from parameter expression."""
+        for parameter in param_expr.parameters:
+            self.parameters.add(parameter)
+
+
+class ParameterManager:
```

Comment:
This whole class looks great! All that logic that was strewn around, now in one place!